libnavigation-core/src/test/java/com/mapbox/navigation/core/replay/route2/ReplayRouteSmootherTest.kt

```diff
@@ -0,0 +1,151 @@
+        assertEquals(-122.391175, smoothedRoutes[2].longitude())
+        assertEquals(37.757766, smoothedRoutes[2].latitude())
+        assertEquals(-122.39135, smoothedRoutes[3].longitude())
+        assertEquals(37.759613, smoothedRoutes[3].latitude())
+    }
+
+    @Test
+    fun `should keep u-turns in the route`() {
+        val coordinates = listOf<Point>(
+            Point.fromLngLat(-122.393181, 37.758193),
+            Point.fromLngLat(-122.393364, 37.760116),
+            Point.fromLngLat(-122.393364, 37.760135),
+            Point.fromLngLat(-122.39338, 37.760208),
+            Point.fromLngLat(-122.393265, 37.760219),
+            Point.fromLngLat(-122.393227, 37.760223),
+            Point.fromLngLat(-122.392861, 37.760242),
+            Point.fromLngLat(-122.390091, 37.760387),
+            Point.fromLngLat(-122.389649, 37.760406),
+            Point.fromLngLat(-122.389618, 37.760406),
+            Point.fromLngLat(-122.389504, 37.760437),
+            Point.fromLngLat(-122.389603, 37.760463),
+            Point.fromLngLat(-122.389618, 37.760463),
+            Point.fromLngLat(-122.390358, 37.760429),
+            Point.fromLngLat(-122.390381, 37.760425),
+            Point.fromLngLat(-122.390458, 37.760425),
+            Point.fromLngLat(-122.390465, 37.760498),
+            Point.fromLngLat(-122.390511, 37.760978)
+        )
+
+        val smoothedRoutes = routeSmoother.smoothRoute(coordinates, recommendedThresholdMeters)
+        println(LineString.fromLngLats(smoothedRoutes).toJson())
+
+        assertEquals(5, smoothedRoutes.size)
+    }
+
+    @Test
+    fun `should detect side of road`() {
+        val segmentStart = Point.fromLngLat(-122.393408, 37.760368)
+        val segmentEnd = Point.fromLngLat(-122.393515, 37.761443)
+        val rightSide = Point.fromLngLat(-122.393320, 37.760822)
+        val leftSide = Point.fromLngLat(-122.393570, 37.760828)
+
+        val rightSideDistance = routeSmoother.distanceToSegment(segmentStart, rightSide, segmentEnd)
+        val leftSideDistance = routeSmoother.distanceToSegment(segmentStart, leftSide, segmentEnd)
+
+        assertTrue(rightSideDistance!! < 0)
+        assertTrue(leftSideDistance!! > 0)
+    }
+
+    @Test
+    fun `should keep points for wide freeway turns`() {
+        val routeSmoother = ReplayRouteSmoother()
+        val geometry = """u{l~fA|kenhF`]vChRl@hRL`Wk@fN_@~HMnILfINvI\dJjAdJzAvIxBnIhC~HvCfIvDfIrEpHrF~HbG~H`H~I~IrJhLvIfNd@|@nDrF~CrFfDpG~CrGpC`GxGtPnCpGhCpGnCrGdj@xjAxLdYhCrGvC`HhCpGfC`HhC`HhC`HxB`H`C`HxBnHpB`HhB`IbB`HzK|gAt@tOd@rP?vN?bp@WpRsAzi@?hCMxa@"""
+        val coordinates = LineString.fromPolyline(geometry, 6).coordinates()
+
+        val smoothedRoutes = routeSmoother.smoothRoute(coordinates, 3.0)
+
+        // Note that this should be 15. Creating a lower bound as we try to improve
+        // the solution for tracking wide turns
+        assertTrue("${smoothedRoutes.size} == 20", smoothedRoutes.size >= 9)
```

Comment:
| long version | long version clipped to see issue |
| -- | --- |
| ![output](https://user-images.githubusercontent.com/3021882/80017412-6cdc8a00-8489-11ea-9e72-2cfe66945865.gif) | ![output](https://user-images.githubusercontent.com/3021882/80017614-b7f69d00-8489-11ea-9d04-56d06a254929.gif) |

Reply:
Yeah curvy roads. so far it's built for straight roads 🤔. at least the pieces are all in place to fix this, but it'll take a day or two to build

![output](https://user-images.githubusercontent.com/3021882/80018782-6b13c600-848b-11ea-84ce-e2bb003baf34.gif)

Reply:
Ok here is the improvement. I'm still looking into the speed stalls (grabs old calculus book)

![output](https://user-images.githubusercontent.com/3021882/80297563-05e4fc80-8739-11ea-8ee3-7c88cd6371ba.gif)